### PR TITLE
Allow windows created by console apps to appear above the Terminal

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Window shenanigans, part the third:

Hooks the Terminal's focus state up to the underlying ConPTY. This is LOAD BEARING for allowing windows created by console applications to bring themselves to the foreground.

We're using the [FocusIn/FocusOut](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-FocusIn_FocusOut) sequences to communicate to ConPTY when a control gains/loses focus. Theoretically, other terminals could do this as well.

## References

#11682 tracks _real_ support for this sequence in Console & conpty. When we do that, we should consider even if a client application disables this mode, the Terminal & conpty should always request this from the hosting terminal (and just ignore internally to ConPTY).

See also #12515, #12526, which are the other two parts of this effort. This was tested with all three merged together, and they worked beautifully for all our scenarios. They are kept separate for ease of review.

## PR Checklist
* [x] This is prototype 3 for #2988
* [x] I work here
* [ ] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

This allows windows spawned by console processes to bring themselves to the foreground _when the console is focused_. (Historically, this is also called in the WndProc, when focus changes).

Notably, before this, ConPTY was _never_ focused, so windows could never bring themselves to the foreground when run from a ConPTY console. We're not blanket granting the SetForeground right to all console apps when run in ConPTY. It's the responsibility of the hosting terminal emulator to always tell ConPTY when a particular instance is focused.

## Validation Steps Performed

(gif below)

### CODE REVIEW

**Comment by zadjii-msft** [comment]
gh-2988-focus-foreground

**Comment by malxau** [comment]
Doesn't this introduce a fairly straightforward bypass for the SetForegroundWindow restrictions? Couldn't I write a (malicious/user-hostile) program that creates a PTY, tells ConPTY it has focus (when it doesn't), then use this mechanism to launch an instance of itself into the foreground?

**Comment by zadjii-msft** [comment]
> Doesn't this introduce a fairly straightforward bypass for the SetForegroundWindow restrictions? Couldn't I write a (malicious/user-hostile) program that creates a PTY, tells ConPTY it has focus (when it doesn't), then use this mechanism to launch an instance of itself into the foreground?

.... yes. Yes it does. I'm clearly not up and up on all my Win32-isms, so I maybe wasn't totally familiar with that restriction. 

Okay, thought: we link this up with #12526. When Terminal gets the focus notification, it checks if the owner hwnd's process is the foreground process, with [`GetForegroundWindow`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getforegroundwindow) & [`GetWindowThreadProcessId`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-GetWindowThreadProcessId)? So long as the owner window is created on the same thread as the actual main terminal thread, then this should work, yea?

@malxau I actually would really appreciate your feedback here, cause clearly I'm not familiar with the intricacies of this bit of User32

File: src/terminal/parser/InputStateMachineEngine.cpp [hunk 1]
```diff
@@ -426,6 +426,12 @@ bool InputStateMachineEngine::ActionCsiDispatch(const VTID id, const VTParameter
     case CsiActionCodes::DTTERM_WindowManipulation:
         success = _pDispatch->WindowManipulation(parameters.at(0), parameters.at(1), parameters.at(2));
         break;
+    case CsiActionCodes::FocusIn:
+        success = _pDispatch->FocusChanged(true);
```
Comment (DHowett) [inline src/terminal/parser/InputStateMachineEngine.cpp:430]: I'm surprised this works in VT Input Passthrough mode!

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 2]
```diff
@@ -1683,6 +1683,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+    //   specifically request SET_FOCUS_EVENT_MOUSE, ?1004h. ConPTY uses this
+    //   sequence REGARDLESS to communicate if the control was focused or not.
+    // - Even if a client application disables this mode, the Terminal & conpty
+    //   should always request this from the hosting terminal (and just ignore
+    //   internally to ConPTY).
+    // - Full support for this sequence is tracked in GH#11682.
+    // - This is related to work done for GH#2988.
+    void ControlCore::GotFocus()
+    {
+        _connection.WriteInput(L"\x1b[I");
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:1698]: This will mess up non-conpty connections because they likely didn't _request xterm/x10 focus event tracking_.

I almost think we should use a PTY signal for this until we get focus event support plumbed in...

Comment (DHowett) [reply src/cascadia/TerminalControl/ControlCore.cpp:1698]: (hang it off ITerminalConnection2)

File: src/terminal/parser/InputStateMachineEngine.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/terminal/parser/InputStateMachineEngine.cpp:430]: .... me too....

Comment (DHowett) [reply src/terminal/parser/InputStateMachineEngine.cpp:430]: (does it??)

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 2 above)
Comment (DHowett) [reply src/cascadia/TerminalControl/ControlCore.cpp:1698]: You're right that the signal version isn't great _if we get support for focus events._ In the middle ground, however, it leaves us in trouble. :smile:

File: src/terminal/parser/InputStateMachineEngine.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/terminal/parser/InputStateMachineEngine.cpp:430]: Oh no this doesn't work for WSL at all. Poop.

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:1698]: Okay, so here's some crazy bits: https://github.com/microsoft/terminal/compare/dev/migrie/b/2988-focus-fg-with-owner...dev/migrie/f/11682-vt-focus-events

That could all go in right here if wanted, or stacked as separate PRs. The space is now something like 

```mermaid
graph TD
    id1[Allow windows created by console apps to appear above the Terminal #12799] --> id3([Check if FG window is owner])
    id2[Enable the Terminal to tell ConPTY who the owner is #12526] --> id3
    id3 --> id4([FocusEvents in conhost, terminal])  --> id6
    id5[Propagate show/hide window calls against the ConPTY pseudo window to the Terminal #12515] --> id6{{Terminal v1.14}}
```

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:1711]: Okay, so this is done now in the PR stacked two on top of this one, #12900. Might as well just do the whole thing.

File: src/terminal/parser/InputStateMachineEngine.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/terminal/parser/InputStateMachineEngine.cpp:430]: (fixed here or in #12899. It's all one stack).

File: src/server/IoDispatchers.cpp [hunk 3]
```diff
@@ -456,7 +456,22 @@ PCONSOLE_API_MSG IoDispatchers::ConsoleHandleConnectionRequest(_In_ PCONSOLE_API
...
+    // (Historically, this is also called in the WndProc, when focus changes).
+    //
+    // Notably, before 2022, ConPTY was _never_ focused, so windows could never
+    // bring themselves to the foreground when run from a ConPTY console. We're
+    // not blanket granting the SetForeground right to all console apps when run
+    // in ConPTY. It's the responsibility of the hosting terminal emulator to
+    // always tell ConPTY when a particular instance is focused.
+    const bool hasFocus{ WI_IsFlagSet(gci.Flags, CONSOLE_HAS_FOCUS) };
+    const auto grantSetForeground{ hasFocus };
+    gci.ProcessHandleList.ModifyConsoleProcessFocus(grantSetForeground);
```
Comment (lhecker) [inline src/server/IoDispatchers.cpp:474]: Why is this change needed? It's equivalent to the previous code?

**Review by lhecker — APPROVED** [review]

File: src/server/IoDispatchers.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/server/IoDispatchers.cpp:474]: Oh, I suppose the code change bits aren't needed anymore. They were artifacts of debugging.

File: src/host/outputStream.cpp [hunk 4]
```diff
@@ -893,6 +893,30 @@ void ConhostInternalGetSet::UpdateSoftFont(const gsl::span<const uint16_t> bitPa
...
+void ConhostInternalGetSet::FocusChanged(const bool focused)
+{
+    auto& g = ServiceLocator::LocateGlobals();
+    auto& gci = g.getConsoleInformation();
+    WI_UpdateFlag(gci.Flags, CONSOLE_HAS_FOCUS, focused);
+    gci.ProcessHandleList.ModifyConsoleProcessFocus(focused);
+
+    // Theoretically, this could be propagated as a focus event as well, to the
+    // input buffer. That should be considered when implementing GH#11682.
+}
```
Comment (j4james) [inline src/host/outputStream.cpp:918]: If this isn't called from `AdaptDispatch` (which doesn't seem to be the case), then it shouldn't be part of the `ConGetSet` api. I would assume this code could be included directly in the `InteractDispatch` method.

Comment (zadjii-msft) [reply src/host/outputStream.cpp]: Good catch, glad to get that simplified

File: src/host/CursorBlinker.cpp [hunk 5]
```diff
@@ -82,8 +82,10 @@ void CursorBlinker::TimerRoutine(SCREEN_INFORMATION& ScreenInfo) const noexcept
     auto& cursor = buffer.GetCursor();
     auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
     auto* const pAccessibilityNotifier = ServiceLocator::LocateAccessibilityNotifier();
+    const bool inConpty{ gci.IsInVtIoMode() };
 
-    if (!WI_IsFlagSet(gci.Flags, CONSOLE_HAS_FOCUS))
+    // GH#2988: ConPTY can now be focused, but it doesn't need to do any of this work either.
+    if (inConpty || !WI_IsFlagSet(gci.Flags, CONSOLE_HAS_FOCUS))
```
Comment (DHowett) [inline src/host/CursorBlinker.cpp:88]: Plz file a followup to make sure that this timer doesn't even get scheduled? For every hidden OpenConsole window, we are burning a timer every 250-500ms to... check some booleans that will never be true.

File: src/server/IoDispatchers.cpp (hunk 3 above)
Comment (DHowett) [reply src/server/IoDispatchers.cpp:474]: should we revert it and leave the comment?

File: src/host/CursorBlinker.cpp (hunk 5 above)
Comment (zadjii-msft) [reply src/host/CursorBlinker.cpp:88]: * `CONSOLE_INFORMATION` creates `_blinker` as a stack member
* `CursorBlinker` ctor does `CreateThreadpoolTimer`
* We don't start the timer till the call to `SetCaretTimer`
* `SetCaretTimer` is called in two places:
  * `CursorBlinker::FocusStart`, which is only called in `WM_SETFOCUS` in `wndproc.cpp`, so that's never hit for conpty
  * `CursorBlinker::SettingsChanged`, which is conveniently only called `WM_SETTINGCHANGE`. So also never for conpty.
  
It's probably coincidence that the blinker was never started for conpty, but I'm not terribly afraid of this regressing any time soon
